types/signed_header_test.go

```diff
@@ -15,89 +16,141 @@ func TestVerify(t *testing.T) {
-				untrustedAdj.Header.BaseHeader.Time = uint64(untrustedAdj.Header.Time().Add(time.Minute).UnixNano())
-				return untrustedAdj
+			prepare: func() (*SignedHeader, bool) {
+				untrusted := *untrustedAdj
+				untrusted.Header.BaseHeader.Time = uint64(untrusted.Header.Time().Add(time.Minute).UnixNano())
+				return &untrusted, true
+			},
+			err: &header.VerifyError{
+				Reason: &ErrNewHeaderTimeFromFuture{},
 			},
-			err: true,
 		},
 		{
-			prepare: func() *SignedHeader {
-				untrustedAdj.BaseHeader.ChainID = "toaster"
-				return untrustedAdj
+			prepare: func() (*SignedHeader, bool) {
+				untrusted := *untrustedAdj
+				untrusted.BaseHeader.ChainID = "toaster"
+				return &untrusted, false // Signature verification should fail
+			},
+			err: &header.VerifyError{
+				Reason: ErrSignatureVerificationFailed,
 			},
-			err: true,
 		},
 		{
-			prepare: func() *SignedHeader {
-				untrustedAdj.Version.App = untrustedAdj.Version.App + 1
-				return untrustedAdj
+			prepare: func() (*SignedHeader, bool) {
+				untrusted := *untrustedAdj
+				untrusted.Version.App = untrusted.Version.App + 1
+				return &untrusted, false // Signature verification should fail
+			},
+			err: &header.VerifyError{
+				Reason: ErrSignatureVerificationFailed,
 			},
-			err: true,
 		},
 		{
-			prepare: func() *SignedHeader {
-				untrustedAdj.ProposerAddress = nil
-				return untrustedAdj
+			prepare: func() (*SignedHeader, bool) {
+				untrusted := *untrustedAdj
+				untrusted.ProposerAddress = nil
+				return &untrusted, true
+			},
+			err: &header.VerifyError{
+				Reason: ErrNoProposerAddress,
 			},
-			err: true,
 		},
 	}
 
 	for i, test := range tests {
 		t.Run(strconv.Itoa(i), func(t *testing.T) {
-			err := trusted.Verify(test.prepare())
-			if test.err {
-				assert.Error(t, err)
-			} else {
+			preparedHeader, recomputeCommit := test.prepare()
+			if recomputeCommit {
+				commit, err := getCommit(preparedHeader.Header, privKey)
+				require.NoError(t, err)
+				preparedHeader.Commit = *commit
+			}
+			err = trusted.Verify(preparedHeader)
+			if test.err == nil {
 				assert.NoError(t, err)
+			} else {
+				if err != nil {
+					switch (err).(type) {
+					case *header.VerifyError:
+						reason := err.(*header.VerifyError).Reason
+						testReason := test.err.(*header.VerifyError).Reason
+						switch reason.(type) {
+						case *ErrLastHeaderHashMismatch:
+							assert.Equal(t, &ErrLastHeaderHashMismatch{}, testReason)
+						case *ErrLastCommitHashMismatch:
+							assert.Equal(t, &ErrLastCommitHashMismatch{}, testReason)
+						case *ErrNewHeaderTimeBeforeOldHeaderTime:
+							assert.Equal(t, &ErrNewHeaderTimeBeforeOldHeaderTime{}, testReason)
+						case *ErrNewHeaderTimeFromFuture:
+							assert.Equal(t, &ErrNewHeaderTimeFromFuture{}, testReason)
+						default:
+							assert.Equal(t, testReason, reason)
+						}
+					default:
+						assert.Failf(t, "unexpected error: %s\n", err.Error())
+					}
+				}
 			}
```

Comment:
I think we are missing a test case, and we can see that better if we focus on de-indenting to handle case by case.
```suggestion
                 // Case 1 & 2: test.err == nil and err == nil or err != nil
	        if test.err == nil {
		        assert.NoError(t, err)
		        return
	        }
	        // Case 3: test.err != nil and err != nil
	        if err == nil {
	          t.Errorf("expected err: %v, got nil",test.err)
	          return
	        }
		// Case 4: test.err != nil and err != nil
	        switch (err).(type) {
	        case *header.VerifyError:
		        reason := err.(*header.VerifyError).Reason
		        testReason := test.err.(*header.VerifyError).Reason
		        switch reason.(type) {
		        case *ErrLastHeaderHashMismatch:
			        assert.Equal(t, &ErrLastHeaderHashMismatch{}, testReason)
		        case *ErrLastCommitHashMismatch:
			        assert.Equal(t, &ErrLastCommitHashMismatch{}, testReason)
		        case *ErrNewHeaderTimeBeforeOldHeaderTime:
			        assert.Equal(t, &ErrNewHeaderTimeBeforeOldHeaderTime{}, testReason)
		        case *ErrNewHeaderTimeFromFuture:
			        assert.Equal(t, &ErrNewHeaderTimeFromFuture{}, testReason)
		        default:
			        assert.Equal(t, testReason, reason)
		        }
	        default:
		        assert.Failf(t, "unexpected error: %s\n", err.Error())
	        }
```